### PR TITLE
JIT: Add a unified mechanism to track metadata/metrics about the compilation

### LINKED ISSUE #52877: SuperPMI asm diffs and metrics improvement
This issue proposes adding a new JIT-EE api for the JIT to report "metrics" back to SuperPMI, and how that could be used by SuperPMI replay and asm diffs.

Currently, running SuperPMI asm diffs happens as follows:
1. Run `superpmi.py asmdiffs` to orchestrate the process.
2. It runs `superpmi.exe -a`.
   a. Each of baseline and diff JIT compile a method context and report results back to the "VM" (or here, the SuperPMI host).
   b. superpmi.exe runs the "near differ" to determine if there is a binary diff. If so, the method context number is added to a list of method contexts with asm diffs.
3. `superpmi.py` iterates over the list of method context numbers, invoking both the baseline and diff JIT to generate textual .dasm (and, thus, compiling the functions again).
4. `superpmi.py` calls `jit-analyze` to analyze the results. It parses the textual output and generates improvement/regression information for each metric it is concerned with.

I propose a set of new JIT-EE methods:
```
void reportMetric(const char* metricName, unsigned int metricValue);
void reportMetric(const char* metricName, unsigned __int64 metricValue);
void reportMetric(const char* metricName, double metricValue);
void reportMetric(const char* metricName, const char* metricValue);
```
If we don't want named overloading, we could add the metric type to the function name. Or, we could have something like:
```
struct JitMetricValue
{
    enum class JitMetricType { TypeInt, TypeLong, TypeDouble, TypeString };
    JitMetricType jitMetricType;
    union
    {
        unsigned int intValue; // jitMetricType == JitMetricType::TypeInt
        unsigned __int64 longValue; // jitMetricType == JitMetricType::TypeLong
        double doubleValue; // jitMetricType == JitMetricType::TypeDouble
        const char* stringValue; // jitMetricType == JitMetricType::TypeString
    };
};
void reportMetric(const char* metricName, JitMetricValue metricValue);
```

At the end of compilation, the JIT calls this API for each interesting metric it has computed. This should be the same set that `jit-analyze` will parse (or a superset), namely, "CodeSize, PerfScore, PrologSize, InstrCount, AllocSize, ExtraAllocBytes, DebugClauseCount, DebugVarCount". For example:
```
reportMetric("CodeSize", emitTotalCodeSize);
```
(we would also report hot/cold/data sizes).

SuperPMI will save these in a new CompileResult table (or tables). The VM would implement this function a no-op, and in fact, the JIT could optionally skip reporting them if SuperPMI is not the host.

For asm diffs, if there is a diff, instead of reporting just the method context#, superpmi.exe could also report the metrics. This could be as a file in CSV form, one column per metric (we might need to handle the case where different compilations report different metrics, in which case maybe json/xml would be better?).

Then, jit-analyze could work directly on this set of data, instead of needing to parse out the metrics from the generated text asm files. This would make jit-analyze much faster. In fact, the set of asm files we generate could be deferred until after jit-analyze is run (not before, as a prerequisite for running jit-analyze). For example, we could choose to only generate dasm files for the set of improved/regressed routines that jit-analyze reports, and not automatically generate dasm for all the method contexts with diffs. In the worst case where every function has diffs, this could save an enormous amount of time, avoiding generating dasm files for thousands of functions which will never be examined. We would need to make sure jit-analyze reports all the diffs that we as developers prefer to look at, including the simplest possible smallest diffs, not just the most impactful diffs. There would need to be a way to have the current behavior of generating all diffs, or perhaps a way to take the list of diffs and generate dasm from them, after the normal processing. Generating the .dasm files could also have a default reasonable cut-off (maximum number of generated files), since generating the dasm is not required to gather the stats.

This depends on all the interesting metrics being available during "normal", non-disasm compilation, which should be true.

So, the new process would look like:
1. Run `superpmi.py asmdiffs` to orchestrate the process.
2. It runs `superpmi.exe -a`.
   a. Each of baseline and diff JIT compile a method context and report results back to the SuperPMI host, _including metrics_.
   b. superpmi.exe runs the "near differ" to determine if there is a binary diff. If so, the method context number is added to a list of method contexts with asm diffs, _including metrics_.
3. `superpmi.py` calls jit-analyze to analyze the results. It reads the generated list of metrics for the difs and generates improvement/regression information for each metric it is concerned with. It needs to generate a machine-readable summary of its results, including method context numbers, in addition to the text/markdown format (there is current provision for TSV/JSON output).
4. `superpmi.py` iterates over the list of method context numbers with diffs that jit-analyze says is interesting, invoking both the baseline and diff JIT to generate textual .dasm for developers to examine.

The work for this should concurrently consider fixing https://github.com/dotnet/runtime/issues/41653, by passing summary of metrics available to superpmi.exe on to jit-analyze. In that case, by somehow providing complete aggregate code size (not just total code size of functions with diffs) to jit-analyze.

@dotnet/jit-contrib Comments?

category:eng-sys
theme:super-pmi
skill-level:expert
cost:large
impact:large

### PR DESCRIPTION
**Pull request by jakobbotsch**
Adds a JIT-EE API to report metadata back to the EE side about the JIT compilation, and adds support for saving these metrics as part of SPMI runs. Switches a number of adhoc metrics to use this scheme, and also adds a few new ones.

The metadata is currently only reported with checked JITs.

Also adds support for fast perfscore diffs, and adds perfscore into the reports generated by `superpmi.py asmdiffs`.

Fix #52877

As future work we should include the metrics as part of the generated diffs report. I think we can do that automatically with relatively little effort. We also may want some variant of `superpmi.py replay` which just collects the metrics into a report.

The system also reports back the method full name and the tiering name as metadata. We can use this in the future to improve the analysis (in particular, grouping repeated diffs in methods of the same full name + tiering name).

For Andy's use case we may consider reporting the metrics back to the EE even in release builds (maybe only under `JitMetrics` or something else that SPMI could set).

### CODE REVIEW

File: src/coreclr/pal/src/safecrt/vsprintf.cpp [hunk 1]
```diff
@@ -95,7 +95,7 @@ DLLEXPORT int __cdecl _vsnprintf_s (
         retvalue = vsnprintf(string, sizeInBytes, format, ap);
         string[sizeInBytes - 1] = '\0';
         /* we allow truncation if count == _TRUNCATE */
-        if (retvalue > (int)sizeInBytes && count == _TRUNCATE)
+        if (retvalue >= (int)sizeInBytes && count == _TRUNCATE)
```
Comment (jakobbotsch) [inline src/coreclr/pal/src/safecrt/vsprintf.cpp:98]: Will submit this fix separately... it should have a unit test added as well.

Comment (jakobbotsch) [reply src/coreclr/pal/src/safecrt/vsprintf.cpp:98]: #98687

File: src/coreclr/jit/jitmetadatalist.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+JITMETADATAINFO(TieringName,                   const char*,      0)
+JITMETADATAMETRIC(PhysicallyPromotedFields,    int,              0)
+JITMETADATAMETRIC(LoopsFoundDuringOpts,        int,              0)
+JITMETADATAMETRIC(LoopsCloned,                 int,              0)
+JITMETADATAMETRIC(LoopsUnrolled,               int,              0)
+JITMETADATAMETRIC(LoopAlignmentCandidates,     int,              0)
+JITMETADATAMETRIC(LoopsAligned,                int,              0)
+JITMETADATAMETRIC(VarsInSsa,                   int,              0)
+JITMETADATAMETRIC(HoistedExpressions,          int,              0)
+JITMETADATAMETRIC(RedundantBranchesEliminated, int,              JIT_METADATA_HIGHER_IS_BETTER)
```
Comment (BruceForstall) [inline src/coreclr/jit/jitmetadatalist.h:26]: Are the flags used anywhere? Where is `JIT_METADATA_HIGHER_IS_BETTER` defined? It seems odd to include this file in superpmi source code but not have this defined there.

File: src/coreclr/jit/jitmetadatalist.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+JITMETADATAMETRIC(JumpThreadingsPerformed,     int,              JIT_METADATA_HIGHER_IS_BETTER)
+JITMETADATAMETRIC(Cses,                        int,              0)
```
Comment (BruceForstall) [inline src/coreclr/jit/jitmetadatalist.h:28]: nit, suggestion:
```suggestion
JITMETADATAMETRIC(CseCount,                        int,              0)
```

File: src/coreclr/jit/jitmetadatalist.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+#define JITMETADATAMETRIC(name, type, flags) JITMETADATA(name, type, flags)
+#endif
+
+#if !defined(JITMETADATAINFO) || !defined(JITMETADATAMETRIC)
+#error Define JITMETADATAINFO and JITMETADATAMETRIC before including this file.
+#endif
+
+//              Name,                          type              flags
+JITMETADATAINFO(MethodFullName,                const char*,      0)
+JITMETADATAINFO(TieringName,                   const char*,      0)
```
Comment (BruceForstall) [inline src/coreclr/jit/jitmetadatalist.h:17]: What is the difference between `JITMETADATAINFO` and `JITMETADATAMETRIC`? (Document above).

Also, the answers to MethodFullName and TieringName are hard coded in SPMI playback. Should that be noted somewhere?

File: src/coreclr/jit/jitmetadata.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+// JitMetadata::report: Report metadata back to the EE.
+//
+// Parameters:
+//   comp - Compiler instance
+//   name - Enum entry specifying the metadata name
+//   data - Pointer to the value to report back
+//
+void JitMetadata::report(Compiler* comp, JitMetadataName name, const void* data)
+{
+    comp->info.compCompHnd->reportMetadata(getName(name), data);
```
Comment (BruceForstall) [inline src/coreclr/jit/jitmetadata.cpp:43]: Should the `report` functions do nothing if `RunningSuperPmiReplay()` is false? (I guess the VM does nothing anyway)

**Review by BruceForstall — APPROVED** [review]
LGTM. Some questions/suggestions.

One additional question: should the metadata info list / schema be statically known (defined in a header file), or dynamic? (clients don't know all the metadata names/types, except maybe by convention, beforehand)

**Review by AndyAyersMS — APPROVED** [review]
Looks good, Bruce has already covered anything I'd have commented on.

I would like to update JitMetrics to use a variant of this, but there are some vector-valued things there. Hmm.

File: src/coreclr/jit/jitmetadatalist.h (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/jitmetadatalist.h:26]: They aren't yet, but my plan was to use this to automatically colorize the metrics in the report (once I add that support). Indeed I haven't defined the enum anywhere yet (and it'll probably only end up being define in superpmi and not the JIT).

File: src/coreclr/jit/jitmetadata.cpp (hunk 5 above)
Comment (jakobbotsch) [reply src/coreclr/jit/jitmetadata.cpp:43]: Yeah I figured that we should just let it be up to the EE. If the EE wants to do something with the information then that's up to them. (Also, we don't check for SPMI in release JITs)

File: src/coreclr/jit/jitmetadatalist.h (hunk 4 above)
Comment (jakobbotsch) [reply src/coreclr/jit/jitmetadatalist.h:17]: Added this comment above:
```cpp
// List of metadata that the JIT can report. There are two categories:
//
// - JITMETADATAINFO: General info that can be of any type and that cannot be
//   aggregated in straightforward ways. These properties are not handled
//   automatically; the JIT must explicitly report them using
//   JitMetadata::report, and the SPMI side needs to manually handle (or ignore)
//   them in ICorJitInfo::reportMetadata.
//
// - JITMETADATAMETRIC: Metrics which are numeric types (currently int, double
//   and int64_t types supported). Their reporting is handled automatically and
//   they will be propagated all the way into SPMI replay/diff results.
```

File: src/coreclr/jit/jitmetadatalist.h (hunk 3 above)
Comment (jakobbotsch) [reply src/coreclr/jit/jitmetadatalist.h:28]: Sounds better, renamed it.

**Comment by jakobbotsch** [comment]
> One additional question: should the metadata info list / schema be statically known (defined in a header file), or dynamic? (clients don't know all the metadata names/types, except maybe by convention, beforehand)

I really want the SPMI side to be able to know what metrics to expect beforehand, just because it simplifies things in the csv handling and also allows e.g. superpmi.py to query metadata about the metrics separately, like the "higher is better" info. There may also be other kinds of metadata in the future (for example, some metrics make sense to aggregate as sums/averages, while some other metrics may make more sense to aggregate as min/max). I'd like to be able to have one place to define all of this and automatically have it propagate all the way into superpmi.py -- and I was trying to make jitmetadatalist.h that place.

Initially I was going to add a new method to the host interface to be able to get the information about the metrics dynamically. For example, then the JIT side would be the only place we used jitmetadatalist.h, to report this information back. It would probably be slightly more flexible than the current approach, which has some downsides. For example, in the current approach if you change the type of a metric then running superpmi.exe with an old JIT will make it misinterpret the metric data as the new type when the old JIT is reports the old type. However, it seemed much easier to just use jitmetadatalist.h from SPMI always, and I would expect the problematic changes to jitmetadatalist.h to be rare (and if we need to make them we always have the hammer of updating the JIT-EE GUID).

> I would like to update JitMetrics to use a variant of this, but there are some vector-valued things there. Hmm.

I added a `size_t length` arg to the API. That seems like good API hygiene regardless when dealing with `void*`, and you can use it for your case to report the full array of metadata in one shot (although you're still going to need some manual handling on the SPMI side, but I guess that's inevitable).

**Pull request merged**